### ISSUE TITLE
Potential bug when using `Meta.auto_assign_tags` and `Meta.raise_on_unknown_json_key` together

### ISSUE THREAD

**Issue by leifdenby** [rnag/dataclass-wizard#137]
* Dataclass Wizard version: `0.22.3`
* Python version: `3.12.4`
* Operating System: macos (`Darwin CM-F16RXCV0WY 23.3.0 Darwin Kernel Version 23.3.0: Wed Dec 20 21:30:59 PST 2023; root:xnu-10002.81.5~7/RELEASE_ARM64_T6030 arm64`)

### Description

I am trying to use `Meta.auto_assign_tags` and `Meta.raise_on_unknown_json_key` together so that I can both define exactly which type should be use when deserializing from `Union` types in a dataclass (as suggested in ) and also use `Meta.raise_on_unknown_json_key = True` so that I can ensure that the parsed json doesn't contain any keys that it shouldn't. However, it appears to me that when using `Meta.auto_assign_tags = True` then the `__tag__` key that is set in the serialised json content is picked up as being an "unknown key". This seems strange to me as it would be nice to use these two together. Even more strangely I have found out that if I create another dataclass where I don't set the `Meta.raise_on_unknown_json_key` to `True` then somehow the issue goes away...

Maybe I am doing something incorrectly, but I am very confused 😕 I am very grateful for any help!

### What I Did

Here's short code snippet the demonstrates what I mean (to duplicate the behaviour where the issue goes away change `CREATE_CONTAINER1` to `True`):

```python
from dataclasses import dataclass
from dataclass_wizard import JSONWizard, YAMLWizard
from typing import Union

@dataclass
class A:
    mynumber: int

@dataclass
class B:
    mystring: str

CREATE_CONTAINER1 = False

if CREATE_CONTAINER1:
    @dataclass
    class Container1(JSONWizard, YAMLWizard):
        obj1: Union[A, B]

        class _(JSONWizard.Meta):
            auto_assign_tags = True


@dataclass
class Container2(JSONWizard, YAMLWizard):
    obj2: Union[A, B]

    class _(JSONWizard.Meta):
        auto_assign_tags = True
        raise_on_unknown_json_key = True

if CREATE_CONTAINER1:
    c1 = Container1(obj1=B("foo"))
    assert c1 == Container1.from_json(c1.to_json())
    print(c1.to_json(indent=2))

c2 = Container2(obj2=B("bar"))
print(c2.to_json(indent=2))
assert c2 == Container2.from_json(c2.to_json())

if CREATE_CONTAINER1:
    Container1.from_dict({"obj1": {"mystring": "bar", "__tag__": "B"}, "bar": 42})

Container2.from_dict({"obj2": {"mystring": "bar", "__tag__": "B"}})   
```

With the above example if I set `CREATE_CONTAINER1` to `False` the above code crashes with:

```python
Traceback (most recent call last):
  File "/Users/B280936/git-repos/mllam/neural-lam/dw_bug.py", line 43, in <module>
    assert c2 == Container2.from_json(c2.to_json())
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/B280936/git-repos/mllam/neural-lam/.venv/lib/python3.12/site-packages/dataclass_wizard/serial_json.py", line 48, in from_json
    return fromdict(cls, o) if isinstance(o, dict) else fromlist(cls, o)
           ^^^^^^^^^^^^^^^^
  File "/Users/B280936/git-repos/mllam/neural-lam/.venv/lib/python3.12/site-packages/dataclass_wizard/loaders.py", line 537, in fromdict
    return load(d)
           ^^^^^^^
  File "/Users/B280936/git-repos/mllam/neural-lam/.venv/lib/python3.12/site-packages/dataclass_wizard/loaders.py", line 620, in cls_fromdict
    cls_kwargs[field_name] = field_to_parser[field_name](
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/B280936/git-repos/mllam/neural-lam/.venv/lib/python3.12/site-packages/dataclass_wizard/parsers.py", line 241, in __call__
    return self.tag_to_parser[tag](o)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/B280936/git-repos/mllam/neural-lam/.venv/lib/python3.12/site-packages/dataclass_wizard/parsers.py", line 72, in __call__
    return self.hook(o, self.base_type)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/B280936/git-repos/mllam/neural-lam/.venv/lib/python3.12/site-packages/dataclass_wizard/loaders.py", line 612, in cls_fromdict
    field_name = lookup_field_for_json_key(o, json_key)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/B280936/git-repos/mllam/neural-lam/.venv/lib/python3.12/site-packages/dataclass_wizard/loaders.py", line 706, in lookup_field_for_json_key
    raise e from None
dataclass_wizard.errors.UnknownJSONKey: A JSON key is missing from the dataclass schema for class `B`.
  unknown key: '__tag__'
  dataclass fields: ['mystring']
  input JSON object: {"mystring": "bar", "__tag__": "B"}
```

However strangely if I define the other dataclass first (the one where `raise_on_unknown_json_key` is not set and therefore defaults to `False`) by changing `CREATE_CONTAINER1` to `True` then the code runs just fine.

**Comment by rnag** [comment]
@leifdenby Thanks for opening this issue. I'll look into this shortly. I haven't tested the code, but it's quite likely this is indeed a bug in the code, when these two `Meta` config values are interleaved together.

**Comment by leifdenby** [comment]
> @leifdenby Thanks for opening this issue! I'll look into this shortly. I haven't tested the code, but it's quite likely this is indeed a bug in the code, when these two `Meta` config values are interleaved together.

Thank you! Let me know if there is anything I can do to help. Also, thank you for making this piece of code. It is so great to have an easy way to go from dataclasses to/from yaml strings. I am using it for configuration in https://github.com/mllam/neural-lam and https://github.com/mllam/mllam-data-prep

**Comment by rnag** [comment]
> > @leifdenby Thanks for opening this issue! I'll look into this shortly. I haven't tested the code, but it's quite likely this is indeed a bug in the code, when these two `Meta` config values are interleaved together.
> 
> Thank you! Let me know if there is anything I can do to help. Also, thank you for making this piece of code. It is so great to have an easy way to go from dataclasses to/from yaml strings. I am using it for configuration in https://github.com/mllam/neural-lam and https://github.com/mllam/mllam-data-prep

No problem! I am glad to hear that this library has helped you 🤗 

I've just released `v0.29.2` and incorporated a bugfix for this issue - check it out! I've added an example below, confirming that the issue seems resolved:

```python3
from __future__ import annotations

from dataclasses import dataclass

from dataclass_wizard import JSONWizard
from dataclass_wizard.errors import UnknownJSONKey


@dataclass
class A:
    mynumber: int


@dataclass
class B:
    mystring: str


@dataclass
class Container(JSONWizard):
    obj2: A | B

    class _(JSONWizard.Meta):
        auto_assign_tags = True
        raise_on_unknown_json_key = True


c = Container(obj2=B("bar"))

output_dict = c.to_dict()

assert output_dict == {
    "obj2": {
        "mystring": "bar",
        "__tag__": "B"
    }
}

assert c == Container.from_dict(output_dict)

# Confirm `raise_on_unknown_json_key` does work
output_dict['obj2']['extra_key'] = True
try:
    Container.from_dict(output_dict)
except UnknownJSONKey:
    print('Success!')
else:
    print('Failed!')
```

**Comment by leifdenby** [comment]
Thank you @rnag! And for doing this so quick. I will give this a try and let you know how I get on.
